petsc build error after update to hdf5 1.12

### Spack version
Current `develop` branch.

### Steps to reproduce the issue

```console
$ spack install petsc@3.7.7 ^superlu_dist@5.1.3_
```
(The `superlu_dist` version is required until #15551 is merged, an related issue.)

leads to:
```console
/gpfs/alpine/fus123/proj-shared/kaig1/spack/opt/spack/linux-rhel7-power9le/gcc-8.1.1/hdf5-1.12.0-env7kt2iczm4znpvefw4cqb6pbi767a3/include/H5version.h:746:23: error: too few arguments to function 'H5Oget_info3'
   #define H5Oget_info H5Oget_info3
                       ^~~~~~~~~~~~
/tmp/kaig1/spack-stage/spack-stage-petsc-3.7.7-v2q6fsmywzszwwjov2nog6fppzwujepz/spack-src/include/petscviewerhdf5.h:34:37: note: in definition of macro 'PetscStackCallHDF5'
     PetscStackPush(#func);_status = func args;PetscStackPop; if (_status) SETERRQ2(PETSC_COMM_SELF,PETSC_ERR_LIB,"Error in HDF5 call %s() Status %d",#func,(int)_status); \
                                     ^~~~
/tmp/kaig1/spack-stage/spack-stage-petsc-3.7.7-v2q6fsmywzszwwjov2nog6fppzwujepz/spack-src/src/sys/classes/viewer/impls/hdf5/hdf5v.c:729:24: note: in expansion of macro 'H5Oget_info'
     PetscStackCallHDF5(H5Oget_info,(obj, &info));
                        ^~~~~~~~~~~
```

This happens because of an incompatible API change in HDF5 1.12. It affects other packages, too, e.g., #15252.

I can see various options to work around this, ie., limiting hdf5 to :1.10.99, adding `-DH5_USE_110_API`, or updating the code to work with the new API, which in particular for old versions of petsc doesn't seem desirable.

@balay

--- a/var/spack/repos/builtin/packages/petsc/package.py
+++ b/var/spack/repos/builtin/packages/petsc/package.py
@@ -146,7 +146,7 @@ class Petsc(Package):
     depends_on('metis@5:~int64', when='@3.8:+metis~int64')
     depends_on('metis@5:+int64', when='@3.8:+metis+int64')
 
-    depends_on('hdf5+mpi+hl+fortran', when='+hdf5+mpi')
+    depends_on('hdf5@:1.10.99+mpi+hl+fortran', when='+hdf5+mpi')
     depends_on('zlib', when='+hdf5')
     depends_on('parmetis', when='+metis+mpi')
     depends_on('valgrind', when='+valgrind')

--- a/var/spack/repos/builtin/packages/petsc/package.py
+++ b/var/spack/repos/builtin/packages/petsc/package.py
@@ -24,6 +24,7 @@ class Petsc(Package):
     version('develop', branch='master')
     version('xsdk-0.2.0', tag='xsdk-0.2.0')
 
+    version('3.12.4', sha256='56a941130da93bbacb3cfa74dcacea1e3cd8e36a0341f9ced09977b1457084c3')
     version('3.12.3', sha256='91f77d7b0f54056f085b9e27938922db3d9bb1734a2e2a6d26f43d3e6c0cf631')
     version('3.12.2', sha256='d874b2e198c4cb73551c2eca1d2c5d27da710be4d00517adb8f9eb3d6d0375e8')
     version('3.12.1', sha256='b72d895d0f4a79acb13ebc782b47b26d10d4e5706d399f533afcd5b3dba13737')

